### PR TITLE
tests,packaging: add package build support for openSUSE

### PR DESCRIPTION
**Pull request by morphis**
This is currently based on https://github.com/snapcore/snapd/pull/3365 so you will see the changes for Fedora here too until that PR is merged.

### CODE REVIEW

File: packaging/fedora/snap-mgmt.sh [hunk 1]
```diff
@@ -18,7 +32,7 @@ systemctl_stop() {
     fi
 }
 
-if [ "$1" = "purge" ]; then
+purge() {
```
Comment (mvo5) [inline packaging/fedora/snap-mgmt.sh]: I added a forum topic that we need to look into sharing code between this and the debian postrm (https://forum.snapcraft.io/t/share-code-between-debian-postrm-purge-and-snap-mgmt-sh/915)

File: spread.yaml [hunk 2]
```diff
@@ -261,9 +268,20 @@ prepare: |
     # (needs to be in spread.yaml directly because there's nothing else on the filesystem yet)
     if [ -f current.delta ]; then
         tf=$(mktemp)
-        # poor-man's "quiet"
-        apt-get update >& "$tf" || ( cat "$tf"; exit 1 )
-        apt-get install -y xdelta3 curl >& "$tf" || ( cat "$tf"; exit 1 )
+        # NOTE: We can't use tests/lib/pkgdb.sh here as it doesn't exist at
```
Comment (mvo5) [inline spread.yaml]: Thanks for this comment! I was wondering exactly this and you answered it :-D

File: tests/lib/apt.sh [hunk 3]
```diff
@@ -13,6 +13,28 @@ install_build_snapd(){
         mv sources.list.back /etc/apt/sources.list
         apt update
     else
-        distro_install_local_package "$GOHOME"/snapd_*.deb
+        packages=
+        case "$SPREAD_SYSTEM" in
```
Comment (mvo5) [inline tests/lib/apt.sh]: Idle wondering if the name for this file (`apt.sh`)  is still appropriate - maybe it should just get merged with pkgdb.sh - but that is a different PR I guess :)

File: tests/lib/pkgdb.sh [hunk 4]
```diff
@@ -104,6 +156,12 @@ distro_update_package_db() {
         ubuntu-*|debian-*)
             quiet apt-get update
             ;;
+        fedora-*)
+            dnf -y -q upgrade
+            ;;
+        opensuse-*)
+            zypper -q update -y
```
Comment (mvo5) [inline tests/lib/pkgdb.sh]: Funny how similar all the tools are in their CLI interface. I guess apt had a lot of influence here.

File: tests/lib/prepare.sh [hunk 5]
```diff
@@ -96,29 +101,53 @@ prepare_classic() {
     if snap --version |MATCH unknown; then
         echo "Package build incorrect, 'snap --version' mentions 'unknown'"
         snap --version
-        apt-cache policy snapd
+        case "$SPREAD_SYSTEM" in
```
Comment (mvo5) [inline tests/lib/prepare.sh]: Could we abstract this to pkgdb.sh as well maybe? Feels a bit ugly to have this test here.

File: packaging/fedora/snap-mgmt.sh (hunk 1 above)
Comment (morphis) [inline packaging/fedora/snap-mgmt.sh]: Thanks!

File: spread.yaml (hunk 2 above)
Comment (morphis) [inline spread.yaml]: :-)

File: tests/lib/prepare.sh (hunk 5 above)
Comment (morphis) [inline tests/lib/prepare.sh]: Done.

File: tests/lib/apt.sh (hunk 3 above)
Comment (morphis) [inline tests/lib/apt.sh]: Done

File: tests/main/ack/task.yaml [hunk 6]
```diff
@@ -1,5 +1,4 @@
 summary: Check snap ack
-systems: [-ubuntu-core-16-arm-*]
```
Comment (morphis) [inline tests/main/ack/task.yaml]: Hm, that shouldn't happen. Let me fix that.

File: packaging/opensuse-42.2/snapd.spec [hunk 7]
```diff
@@ -171,7 +173,7 @@ install -m 644 -D packaging/opensuse-42.2/permissions %buildroot/%{_sysconfdir}/
 install -m 644 -D packaging/opensuse-42.2/permissions.paranoid %buildroot/%{_sysconfdir}/permissions.d/snapd.paranoid
 # Install the systemd units
 make -C data/systemd install DESTDIR=%{buildroot} SYSTEMDSYSTEMUNITDIR=%{_unitdir}
-for s in snapd.autoimport.service snapd.system-shutdown.service; do
+for s in snapd.autoimport.service snapd.system-shutdown.service snap-repair.timer snap-repair.service; do
```
Comment (zyga) [inline packaging/opensuse-42.2/snapd.spec]: I was under the impression that the repair functionality is only applicable to core systems. I would not recommend using it in any classic distribution where existing packaging systems are sufficient to recover from any issue.

File: spread.yaml [hunk 8]
```diff
@@ -317,7 +322,7 @@ restore: |
 suites:
     tests/main/:
         summary: Full-system tests for snapd
-        systems: [-fedora-*]
+        systems: [-fedora-*, -opensuse-*]
```
Comment (zyga) [inline spread.yaml]: Can you add a comment explaining why those are disabled?

File: spread.yaml [hunk 9]
```diff
@@ -336,14 +341,17 @@ suites:
             if [[ "$SPREAD_SYSTEM" != ubuntu-core-16-* ]]; then
                 . $TESTSLIB/pkgdb.sh
                 distro_purge_package snapd
-                distro_purge_package snap-confine
+                if [[ "$SPREAD_SYSTEM" != opensuse-* ]]; then
```
Comment (zyga) [inline spread.yaml]: This is where I would argue that the following solution is cleaner:
```
# NOTE: snap-confine never existed on openSUSE
distro-purge-package {debian,fedora}:snap-confine`
```

File: spread.yaml [hunk 10]
```diff
@@ -336,14 +341,17 @@ suites:
...
+                    distro_purge_package snap-confine
+                fi
             fi
 
     tests/completion/:
         summary: completion tests
 
         # ppc64el disabled because of https://bugs.launchpad.net/snappy/+bug/1655594
-        systems: [-ubuntu-core-*, -ubuntu-*-ppc64el, -fedora-*]
+        systems: [-ubuntu-core-*, -ubuntu-*-ppc64el, -fedora-*, -opensuse-*]
```
Comment (zyga) [inline spread.yaml]: Why are those tests disabled on Fedora and openSUSE? Can you please add a comment.

File: spread.yaml [hunk 11]
```diff
@@ -373,7 +381,7 @@ suites:
 
     tests/regression/:
         summary: Regression tests for snapd
-        systems: [-fedora-*]
+        systems: [-fedora-*, -opensuse-*]
```
Comment (zyga) [inline spread.yaml]: Can you please document why those are disabled? Ideally we'd run them and skip the parts that involve confinement.

File: spread.yaml [hunk 12]
```diff
@@ -392,7 +400,7 @@ suites:
 
     tests/upgrade/:
         summary: Tests for snapd upgrade
-        systems: [-ubuntu-core-16-*, -fedora-*]
+        systems: [-ubuntu-core-16-*, -fedora-*, -opensuse-*]
```
Comment (zyga) [inline spread.yaml]: Can you please document those too?

File: spread.yaml [hunk 13]
```diff
@@ -408,7 +416,7 @@ suites:
 
     tests/unit/:
         summary: Suite to run unit tests (non-go and different go runtimes)
-        systems: [-ubuntu-core-16-*, -fedora-*]
+        systems: [-ubuntu-core-16-*, -fedora-*, -opensuse-*]
```
Comment (zyga) [inline spread.yaml]: Can you please document those too? I thought that we can run unit tests on openSUSE now, is this not the case?

File: spread.yaml [hunk 14]
```diff
@@ -434,7 +442,7 @@ suites:
     tests/nightly/:
         summary: Suite for nightly, expensive, tests
         manual: true
-        systems: [-fedora-*]
+        systems: [-fedora-*, -opensuse-*]
```
Comment (zyga) [inline spread.yaml]: Can you please document those too?

File: spread.yaml [hunk 15]
```diff
@@ -453,7 +461,7 @@ suites:
 
     tests/nested/:
         summary: Tests for nested images
-        systems: [-fedora-*]
+        systems: [-fedora-*, -opensuse-*]
```
Comment (zyga) [inline spread.yaml]: Can you please document those too?

File: tests/lib/pkgdb.sh [hunk 16]
```diff
@@ -31,13 +31,24 @@ fedora_name_package() {
     esac
 }
 
+opensuse_name_package() {
```
Comment (zyga) [inline tests/lib/pkgdb.sh]: Yeah, I also wonder why those function names are backwards.

File: tests/lib/pkgdb.sh [hunk 17]
```diff
@@ -166,13 +192,65 @@ distro_auto_remove_packages() {
...
+        mv sources.list.back /etc/apt/sources.list
+        apt update
+    else
+        packages=
+        case "$SPREAD_SYSTEM" in
+            ubuntu-*|debian-*)
+                packages="${GOHOME}/snapd_*.deb"
+                ;;
+            fedora-*|opensuse-*)
+                packages="${GOHOME}/snap-confine*.rpm ${GOPATH}/snapd*.rpm"
```
Comment (zyga) [inline tests/lib/pkgdb.sh]: snap-confine is not a package for openSUSE, I think you have to split this

File: tests/lib/pkgdb.sh [hunk 18]
```diff
@@ -166,13 +192,65 @@ distro_auto_remove_packages() {
...
+            *)
+                exit 1
+                ;;
+        esac
+
+        distro_install_local_package $packages
+
+        # Perform per distribution post-installation steps if necessary
+        case "$SPREAD_SYSTEM" in
+            opensuse-*)
```
Comment (zyga) [inline tests/lib/pkgdb.sh]: We could query systemd preset policy instead of maintaining a switch statement.

File: tests/lib/prepare-project.sh [hunk 19]
```diff
@@ -66,6 +66,42 @@ fedora_build_rpm() {
     rm $GOPATH/*.src.rpm
 }
 
+opensuse_build_rpm() {
+    release=$(echo "$SPREAD_SYSTEM" | awk '{split($0,a,"-");print a[2]}')
```
Comment (zyga) [inline tests/lib/prepare-project.sh]: Could this just come from `/etc/os-release`? I'm not sure why we're encoding it in SPREAD_SYSTEM

File: tests/lib/prepare-project.sh [hunk 20]
```diff
@@ -66,6 +66,42 @@ fedora_build_rpm() {
...
+    arch=x86_64
+
+    base_version="$(head -1 debian/changelog | awk -F'[()]' '{print $2}')"
+    version="1337.$base_version"
+    sed -i -e "s/^Version:.*$/Version: $version/g" packaging/opensuse-$release/snapd.spec
+
+    mkdir -p /tmp/pkg/snapd-$version
+    cp -rav * /tmp/pkg/snapd-$version/
```
Comment (zyga) [inline tests/lib/prepare-project.sh]: Did you consider using 'osc build' to do a local build of the package?

File: tests/lib/prepare.sh [hunk 21]
```diff
@@ -126,19 +119,29 @@ prepare_classic() {
...
 Environment=SNAPD_DEBUG_HTTP=7 SNAPD_DEBUG=1 SNAPPY_TESTING=1 SNAPD_CONFIGURE_HOOK_TIMEOUT=30s
 EOF
     mkdir -p /etc/systemd/system/snapd.socket.d
     cat <<EOF > /etc/systemd/system/snapd.socket.d/local.conf
 [Unit]
-StartLimitInterval=0
+$START_LIMIT_INTERVAL
 EOF
 
+    systemctl daemon-reload
```
Comment (zyga) [inline tests/lib/prepare.sh]: Can you document why those are invoked here.

File: packaging/opensuse-42.2/snapd.spec (hunk 7 above)
Comment (morphis) [inline packaging/opensuse-42.2/snapd.spec]: That is exactly why we're `rm` these system units here after they got installed by `data/systemd/Makefile` :-)

File: spread.yaml (hunk 8 above)
Comment (morphis) [inline spread.yaml]: Sure.

File: spread.yaml (hunk 9 above)
Comment (morphis) [inline spread.yaml]: It's always a tradeoff. Purging not existing packages is a very rare case where as installing packages is the very regular one. So limiting us to a minimal set of these is enough here.

File: tests/lib/pkgdb.sh (hunk 16 above)
Comment (morphis) [inline tests/lib/pkgdb.sh]: Why are they backwards` It's <distro>_<verb>_<what> were distro=opensuse verb=to name and what=package in this case.

File: tests/lib/pkgdb.sh (hunk 18 above)
Comment (morphis) [inline tests/lib/pkgdb.sh]: Good idea.

File: tests/lib/prepare-project.sh (hunk 20 above)
Comment (morphis) [inline tests/lib/prepare-project.sh]: A `osc build` isn't possible without a login. Also I have code to unite the rpm build for Fedora and openSUSE under a single `build_rpm` function.

File: tests/lib/prepare-project.sh (hunk 19 above)
Comment (morphis) [inline tests/lib/prepare-project.sh]: It's a reliable identifier we own. /etc/os-release isn't really owned by us. Would like to keep these influence out of the picture when we're doing spread specific things like setting up the environment.

File: tests/lib/pkgdb.sh (hunk 17 above)
Comment (morphis) [inline tests/lib/pkgdb.sh]: We could but don't have to, Zypper wont fail interestingly. However let me split these.

File: spread.yaml (hunk 13 above)
Comment (morphis) [inline spread.yaml]: We can run unit tests on openSUSE but as explained before this PR just adds build support for openSUSE and doesn't enable any spread tests (in real we have to enable at least a single one tests/main/help to get the build running). Enabling tests will happen with later PRs step by step.

File: spread.yaml (hunk 11 above)
Comment (morphis) [inline spread.yaml]: Sure, that is what will happen in a later PR.

File: tests/lib/pkgdb.sh (hunk 18 above)
Comment (morphis) [inline tests/lib/pkgdb.sh]: Done

File: tests/lib/pkgdb.sh (hunk 16 above)
Comment (zyga) [inline tests/lib/pkgdb.sh]: It just looks odd as nothing else uses this style. If anything I'd call it opensuse_package_name (`object.object.object` style)

File: packaging/opensuse-42.2/snapd.spec (hunk 7 above)
Comment (zyga) [inline packaging/opensuse-42.2/snapd.spec]: Ohhh, I didn't notice the line below :D

File: tests/main/ack/task.yaml (hunk 6 above)
Comment (zyga) [inline tests/main/ack/task.yaml]: This is still broken

Comment (morphis) [inline tests/main/ack/task.yaml]: Hm, I've pushed a change for that.

**Pull request merged**